### PR TITLE
Implementation multidimensional interpolation on regular grids.

### LINKED ISSUE #382: Multidimensional interpolation on regular grids.
What about implementing multidimensional interpolation on regular grids.

We really need a function similar to the implementation in scipy: **interpn** https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interpn.html

Perhaps this problem can be solved with the existing functionality, please tell me

### PR DESCRIPTION
**Pull request by eisberg**
Implementation multidimensional interpolation on regular grids. 

Closes #382

### CODE REVIEW

**Comment by markrogoyski** [comment]
Hi @eisberg,

Thank you for the pull request for the multivariate regular grid interpolation.

I was able to reproduce your example in the feature request issue ([382](https://github.com/markrogoyski/math-php/issues/382)) as a PHPUnit test, and verify the result using SciPy for both _linear_ and _nearest_ methods. Here is what I came up with based on your example and how I computed the expected result with SciPy.

**PHPUnit Test Cases**
```php
    public function testIssue382ExampleLinear()
    {
        // Given
        $xs = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
        $ys = [10, 11, 12, 13, 14, 15, 16, 17, 18, 19];
        $zs = [110, 111, 112, 113, 114, 115, 116, 117, 118, 119];

        // And
        $func = function ($x, $y, $z)
        {
            return 2 * $x + 3 * $y - $z;
        };

        // And
        $data = [];
        foreach ($xs as $i => $x) {
            foreach ($ys as $j => $y) {
                foreach ($zs as $k => $z) {
                    $data[$i][$j][$k] = $func($x, $y, $z);
                }
            }
        }

        // When
        $interp = new RegularGridInterpolator([$xs, $ys, $zs], $data, 'linear');
        $result = $interp([2.21, 12.1, 115.9]);

        // Then
        $this->assertEquals(-75.18, $result, '', 0.00001);
    }

    public function testIssue382ExampleNearest()
    {
        // Given
        $xs = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
        $ys = [10, 11, 12, 13, 14, 15, 16, 17, 18, 19];
        $zs = [110, 111, 112, 113, 114, 115, 116, 117, 118, 119];

        // And
        $func = function ($x, $y, $z)
        {
            return 2 * $x + 3 * $y - $z;
        };

        // And
        $data = [];
        foreach ($xs as $i => $x) {
            foreach ($ys as $j => $y) {
                foreach ($zs as $k => $z) {
                    $data[$i][$j][$k] = $func($x, $y, $z);
                }
            }
        }

        // When
        $interp = new RegularGridInterpolator([$xs, $ys, $zs], $data, 'nearest');
        $result = $interp([2.21, 12.1, 115.9]);

        // Then
        $this->assertEquals(-76, $result, '', 0.00001);
    }
```

**Python SciPy Expected Result Verification**
```python
>>> xs = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
>>> ys = [10, 11, 12, 13, 14, 15, 16, 17, 18, 19];
>>> zs = [110, 111, 112, 113, 114, 115, 116, 117, 118, 119];

>>> def func(x, y, z):
...   return 2 * x + 3 * y - z

>>> values = [[[func(x, y, z) for z in zs] for y in ys] for x in xs]

>>> my_interpolating_function = RegularGridInterpolator((xs, ys, zs), values, method='linear')
>>> my_interpolating_function([2.21, 12.1, 115.9])
array([-75.18])
>>>
>>>
>>> my_interpolating_function = RegularGridInterpolator((xs, ys, zs), values, method='nearest')
>>> my_interpolating_function([2.21, 12.1, 115.9])
array([-76.])
```

So far it looks good. Give me some time to learn a little more about how this works. If you have more examples to use as unit tests or links to other examples, that would be great.

Thanks again for the pull request. I'll let you know if I have any questions or comments in the next couple days.

Thanks,
Mark

**Comment by eisberg** [comment]
Hi @markrogoyski 

In PR, I added a unit test based on test data and the expected result in the original unit test from Python: `https://github.com/scipy/scipy/blob/c734dacd61c5962a86ab3cc4bf2891fc94b720a6/scipy/interpolate/tests/test_interpolate.py#L2361`:

```python
    def _get_sample_4d(self):
        # create a 4-D grid of 3 points in each dimension
        points = [(0., .5, 1.)] * 4
        values = np.asarray([0., .5, 1.])
        values0 = values[:, np.newaxis, np.newaxis, np.newaxis]
        values1 = values[np.newaxis, :, np.newaxis, np.newaxis]
        values2 = values[np.newaxis, np.newaxis, :, np.newaxis]
        values3 = values[np.newaxis, np.newaxis, np.newaxis, :]
        values = (values0 + values1 * 10 + values2 * 100 + values3 * 1000)
        return points, values
    def test_linear_xi3d(self):
        points, values = self._get_sample_4d()
        interp = RegularGridInterpolator(points, values)
        sample = np.asarray([[0.1, 0.1, 1., .9], [0.2, 0.1, .45, .8],
                             [0.5, 0.5, .5, .5]])
        wanted = np.asarray([1001.1, 846.2, 555.5])
        assert_array_almost_equal(interp(sample), wanted)

    def test_nearest(self):
        points, values = self._get_sample_4d()
        interp = RegularGridInterpolator(points, values, method="nearest")
        sample = np.asarray([0.1, 0.1, .9, .9])
        wanted = 1100.
        assert_array_almost_equal(interp(sample), wanted)
        sample = np.asarray([0.1, 0.1, 0.1, 0.1])
        wanted = 0.
        assert_array_almost_equal(interp(sample), wanted)
        sample = np.asarray([0., 0., 0., 0.])
        wanted = 0.
        assert_array_almost_equal(interp(sample), wanted)
        sample = np.asarray([1., 1., 1., 1.])
        wanted = 1111.
        assert_array_almost_equal(interp(sample), wanted)
        sample = np.asarray([0.1, 0.4, 0.6, 0.9])
        wanted = 1055.
        assert_array_almost_equal(interp(sample), wanted)

    def test_linear_edges(self):
        points, values = self._get_sample_4d()
        interp = RegularGridInterpolator(points, values)
        sample = np.asarray([[0., 0., 0., 0.], [1., 1., 1., 1.]])
        wanted = np.asarray([0., 1111.])
        assert_array_almost_equal(interp(sample), wanted)
```

File: src/NumericalAnalysis/Interpolation/RegularGridInterpolator.php [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+     * Python port np.searchsorted.
+     *
+     * @param $needle
+     *
+     * @return int
+     */
+    private function searchSorted(array $haystack, $needle)
+    {
+        $index = -1;
+        sort($haystack);
```
Comment (markrogoyski) [inline src/NumericalAnalysis/Interpolation/RegularGridInterpolator.php:224]: Doesn't Python's `np.searchsorted` presume that the input array is already sorted? In this case here as well, if you are looking for the insertion point that keeps the order, if you sort the list here, and it wasn't already sorted, the index you find isn't necessarily going to match with what is outside of this back at the caller side, correct?

In Python, if I have a list [1, 2, 3, 4, 5, 9] and I want to insert 8, I will get an index of 5:
```python
>>> np.searchsorted([1, 2, 3, 4, 5, 9], 8)
5
```
However, if the list is not sorted, such as [1, 2, 3, 9, 4, 5], and I want to insert 8, I will get an index of 3:
```python
>>> np.searchsorted([1, 2, 3, 9, 4, 5], 8)
3
```
Python's `np.searchsorted` does not sort the list--it presumes it is already sorted.

Now suppose we have the same array in PHP.
```php
php > $haystack = [1, 2, 3, 9, 4, 5];
php > print_r($haystack);
Array
(
    [0] => 1
    [1] => 2
    [2] => 3
    [3] => 9
    [4] => 4
    [5] => 5
)
```
The 9 is at index 3.
If you sort this in the function, you change the index of where 9 is:
```php
php > sort($haystack);
php > print_r($haystack);
Array
(
    [0] => 1
    [1] => 2
    [2] => 3
    [3] => 4
    [4] => 5
    [5] => 9
)
```

If you sort in the `searchSorted` function and change the indices, but the caller's array still the original indices, the return value of the index isn't going to make sense in the context of the caller with the original indices.

Am I misunderstanding something here? Can you help me better understand the intent of this method and where/if at all things should be sorted?

BTW, if you remove this line `sort($haystack)`, all the unit tests still pass. The haystack array being passed in must already be sorted?

Thanks.

Comment (eisberg) [reply src/NumericalAnalysis/Interpolation/RegularGridInterpolator.php:224]: I think You are right, and sorting can really be removed.

Comment (eisberg) [reply src/NumericalAnalysis/Interpolation/RegularGridInterpolator.php]: It is assumed that the input data will already be sorted

**Comment by markrogoyski** [comment]
Hi @eisberg,

Thank you for the pull request and adding this new functionality. It looks good. This will be included in the next release.

Thanks again.
Mark

**Pull request merged**